File: behavioral/dart/observer_structured/weather_station.dart
import 'interfaces.dart';

class WeatherStation {
  final List<WeatherObserver> _observers = [];
  double _temperature = 0;
  double _humidity = 0;
  double _pressure = 0;
  
  void addObserver(WeatherObserver observer) {
    _observers.add(observer);
  }
  
  void removeObserver(WeatherObserver observer) {
    _observers.remove(observer);
  }
  
  void setMeasurements(double temp, double hum, double press) {
    _temperature = temp;
    _humidity = hum;
    _pressure = press;
    print('\nWeather updated: $_temperature°C, $_humidity% humidity, $_pressure hPa');
    _notifyObservers();
  }
  
  void _notifyObservers() {
    for (final observer in _observers) {
      observer.onWeatherUpdate(_temperature, _humidity, _pressure);
    }
  }
}

class DisplayDevice implements WeatherObserver {
  final String _deviceName;
  
  DisplayDevice(this._deviceName);
  
  @override
  void onWeatherUpdate(double temperature, double humidity, double pressure) {
    print('  [$_deviceName] Displaying: $temperature°C, $humidity%, $pressure hPa');
  }
}

class AlertSystem implements WeatherObserver {
  @override
  void onWeatherUpdate(double temperature, double humidity, double pressure) {
    print('  [AlertSystem] Checking conditions...');
    if (temperature > 35) {
      print('  [AlertSystem] WARNING: High temperature!');
    }
    if (humidity > 80) {
      print('  [AlertSystem] WARNING: High humidity!');
    }
    if (pressure < 980) {
      print('  [AlertSystem] WARNING: Low pressure (storm coming)!');
    }
  }
}

File: behavioral/typescript/observer_structured/weather_station.ts
import { WeatherObserver } from './interfaces';

export class WeatherStation {
  private observers: WeatherObserver[] = [];
  private temperature: number = 0;
  private humidity: number = 0;
  private pressure: number = 0;
  
  addObserver(observer: WeatherObserver): void {
    this.observers.push(observer);
  }
  
  removeObserver(observer: WeatherObserver): void {
    const index = this.observers.indexOf(observer);
    if (index !== -1) {
      this.observers.splice(index, 1);
    }
  }
  
  setMeasurements(temp: number, hum: number, press: number): void {
    this.temperature = temp;
    this.humidity = hum;
    this.pressure = press;
    console.log(`\nWeather updated: ${this.temperature}°C, ${this.humidity}% humidity, ${this.pressure} hPa`);
    this.notifyObservers();
  }
  
  private notifyObservers(): void {
    for (const observer of this.observers) {
      observer.onWeatherUpdate(this.temperature, this.humidity, this.pressure);
    }
  }
}

export class DisplayDevice implements WeatherObserver {
  private deviceName: string;
  
  constructor(name: string) {
    this.deviceName = name;
  }
  
  onWeatherUpdate(temperature: number, humidity: number, pressure: number): void {
    console.log(`  [${this.deviceName}] Displaying: ${temperature}°C, ${humidity}%, ${pressure} hPa`);
  }
}

export class AlertSystem implements WeatherObserver {
  onWeatherUpdate(temperature: number, humidity: number, pressure: number): void {
    console.log('  [AlertSystem] Checking conditions...');
    if (temperature > 35) {
      console.log('  [AlertSystem] WARNING: High temperature!');
    }
    if (humidity > 80) {
      console.log('  [AlertSystem] WARNING: High humidity!');
    }
    if (pressure < 980) {
      console.log('  [AlertSystem] WARNING: Low pressure (storm coming)!');
    }
  }
}
